src/main/java/com/wavefront/generators/FromPatternGenerator.java

```diff
@@ -73,11 +76,25 @@ public FromPatternGenerator(GeneratorConfig config, DataQueue dataQueue) {
   @Override
   public void generateForFile() {
     Function<DataQueue, Boolean> whileCheck = getWhileCheck();
+    long startSeconds = System.currentTimeMillis() / 1000;
+    long currentSeconds = startSeconds;
+    long mustBeGeneratedSpans;
+    int generatedSpans = 0;
+    int rate = generatorConfig.getSpansRate();
     while (whileCheck.apply(dataQueue)) {
-      // get next trace type to be generated
-      TraceTypePattern traceTypePattern = getNextTraceType(generatorConfig.getTraceTypePatterns());
+      // Spread traces across the whole period.
+      // Simulate the delay.
+      currentSeconds += SLEEP_DELAY_SECONDS;
+      mustBeGeneratedSpans = rate * (currentSeconds - startSeconds);
 
-      dataQueue.addTrace(generateTrace(traceTypePattern));
+      while (generatedSpans < mustBeGeneratedSpans) {
+        // get next trace type to be generated
+        TraceTypePattern traceTypePattern = getNextTraceType(generatorConfig.getTraceTypePatterns());
+        Trace trace = generateTrace(traceTypePattern,
+            (currentSeconds - random.nextInt(SLEEP_DELAY_SECONDS)) * 1000);
+        dataQueue.addTrace(trace);
+        generatedSpans += trace.getSpansCount();
+      }
```

Comment:
now this function become very similar to run() implementation, so could you make one parameterized function to not have code duplication.